src/etc/inc/util.inc

```diff
@@ -1134,6 +1134,13 @@ function is_port($port) {
 	return false;
 }
 
+/* returns true if $port is in use */
+function is_port_in_use($port, $proto = "tcp", $ip_version = 4) {
+	exec("/usr/bin/netstat -an " . escapeshellarg('-' . $ip_version) . " -p " . escapeshellarg($proto) . " | /usr/bin/awk '{print $4}' | /usr/bin/egrep -o '*([0-9]{1,5})$' | /usr/bin/sort -u", $port_info);
```

Comment:
`netstat` binary supports libxo(3).  It means it can output in json format which is much better to parse instead of running all these other commands. See the following example and remember to remove `,pretty` from the call since it's not needed here.
```
[2.4.4-RELEASE][root@pfgarga.home]/root: netstat --libxo json,pretty -an -4 -p tcp
{
  "statistics": {
    "socket": [
      {
        "protocol": "tcp4",
        "receive-bytes-waiting": 0,
        "send-bytes-waiting": 44,
        "local": {
          "address": "172.21.4.1",
          "port": "16222"
        },
        "remote": {
          "address": "172.21.4.121",
          "port": "50010"
        },
        "tcp-state": "ESTABLISHED"
      },
      {
        "protocol": "tcp4",
        "receive-bytes-waiting": 0,
        "send-bytes-waiting": 0,
        "local": {
          "address": "127.0.0.1",
          "port": "953"
        },
        "remote": {
          "address": "*",
          "port": "*"
        },
        "tcp-state": "LISTEN"
      },
      {
        "protocol": "tcp4",
        "receive-bytes-waiting": 0,
        "send-bytes-waiting": 0,
        "local": {
          "address": "*",
          "port": "53"
        },
        "remote": {
          "address": "*",
          "port": "*"
        },
        "tcp-state": "LISTEN"
      },
      {
        "protocol": "tcp4",
        "receive-bytes-waiting": 0,
        "send-bytes-waiting": 0,
        "local": {
          "address": "127.0.0.1",
          "port": "19001"
        },
        "remote": {
          "address": "*",
          "port": "*"
        },
        "tcp-state": "LISTEN"
      },
      {
        "protocol": "tcp4",
        "receive-bytes-waiting": 0,
        "send-bytes-waiting": 0,
        "local": {
          "address": "*",
          "port": "80"
        },
        "remote": {
          "address": "*",
          "port": "*"
        },
        "tcp-state": "LISTEN"
      },
      {
        "protocol": "tcp4",
        "receive-bytes-waiting": 0,
        "send-bytes-waiting": 0,
        "local": {
          "address": "*",
          "port": "16443"
        },
        "remote": {
          "address": "*",
          "port": "*"
        },
        "tcp-state": "LISTEN"
      },
      {
        "protocol": "tcp4",
        "receive-bytes-waiting": 0,
        "send-bytes-waiting": 0,
        "local": {
          "address": "127.0.0.1",
          "port": "19000"
        },
        "remote": {
          "address": "*",
          "port": "*"
        },
        "tcp-state": "LISTEN"
      },
      {
        "protocol": "tcp4",
        "receive-bytes-waiting": 0,
        "send-bytes-waiting": 0,
        "local": {
          "address": "*",
          "port": "16222"
        },
        "remote": {
          "address": "*",
          "port": "*"
        },
        "tcp-state": "LISTEN"
      }
    ]
  }
}
```

Reply:
Only downside in this case is that it would require a bit more memory/processing in PHP to handle and parse through the JSON data, but I don't think `netstat` output would ever be large enough for that to be a problem here.

If you want a recent example of how to parse that kind of output, see https://github.com/pfsense/pfsense/blob/master/src/etc/inc/system.inc#L2450 or https://github.com/pfsense/pfsense/blob/master/src/etc/inc/syslog.inc#L722